src/circuit/gadget/utilities/lookup_range_check.rs

```diff
@@ -0,0 +1,297 @@
+            running_sum,
+            table_idx,
+            perm,
+            _marker: PhantomData,
+        };
+
+        meta.lookup(|meta| {
+            let q_lookup = meta.query_fixed(config.q_lookup, Rotation::cur());
+            let z_cur = meta.query_advice(config.running_sum, Rotation::cur());
+            let z_next = meta.query_advice(config.running_sum, Rotation::next());
+            //    z_i = 2^{K}⋅z_{i + 1} + a_i
+            // => a_i = z_i - 2^{K}⋅z_{i + 1}
+            let word = z_cur - z_next * F::from_u64(1 << K);
+            let table = meta.query_fixed(config.table_idx, Rotation::cur());
+
+            vec![(q_lookup * word, table)]
+        });
+
+        config
+    }
+
+    pub fn load(&self, layouter: &mut impl Layouter<F>) -> Result<(), Error> {
+        layouter.assign_region(
+            || "table_idx",
+            |mut gate| {
+                // We generate the row values lazily (we only need them during keygen).
+                for index in 0..(1 << K) {
+                    gate.assign_fixed(
+                        || "table_idx",
+                        self.table_idx,
+                        index,
+                        || Ok(F::from_u64(index as u64)),
+                    )?;
+                }
+                Ok(())
+            },
+        )
+    }
+
+    fn lookup_range_check(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        words: CellValue<F>,
+        num_words: usize,
+    ) -> Result<Vec<CellValue<F>>, Error> {
+        // `num_words` must fit into a single field element.
+        assert!(num_words <= F::NUM_BITS as usize / K);
+        let num_bits = num_words * K;
+
+        // Take first num_bits bits of `words`.
+        let bits = words.value().map(|words| {
+            words
+                .to_le_bits()
+                .into_iter()
+                .take(num_bits)
+                .collect::<Vec<_>>()
+        });
+
+        // Chunk the first num_bits bits into K-bit chunks.
+        let bits: Option<Vec<Vec<bool>>> = bits.map(|bits| {
+            bits.chunks_exact(K)
+                .map(|chunk| chunk.to_vec())
+                .collect::<Vec<_>>()
+        });
+
+        let bits = if let Some(bits) = bits {
+            bits.into_iter().map(Some).collect()
+        } else {
+            vec![None; num_words]
+        };
+
+        // Copy `words` and initialize running sum `z_0 = words` to decompose it.
+        let z_0 = copy(
+            region,
+            || "copy words",
+            self.running_sum,
+            offset,
+            &words,
+            &self.perm,
+        )?;
+
+        let mut zs = vec![z_0];
+
+        // Assign cumulative sum such that
+        //          z_i = 2^{K}⋅z_{i + 1} + a_i
+        // => z_{i + 1} = (z_i - a_i) / (2^K)
+        //
+        // For `words` = a_0 + 2^10 a_1 + ... + 2^{120} a_{12}}, initialize z_0 = `words`.
+        // If `words` fits in 130 bits, we end up with z_{13} = 0.
```

Comment:
Document the `lookup_range_check` function to include this API limitation: only the lower `num_words * K` bits are constrained by this function, and it is up to the caller to constrain `z_{num_words} == 0` (or otherwise constrain the upper bits).